nixos/modules/services/networking/vpnc.nix

```diff
@@ -0,0 +1,97 @@
+{ lib, pkgs, config, ... }:
+
+with lib;
+
+let
+
+  buildVpnc = key: value: {
+    name = "vpnc/${key}.conf";
+    value = {
+      text = ''
+        IPSec gateway ${value.gateway}
+        IPSec ID ${value.id}
+        IPSec secret ${value.secret}
+        Xauth username ${value.username}
+        Password helper ${pkgs.systemd}/bin/systemd-ask-password
+        Script ${pkgs.vpnc}/etc/vpnc/vpnc-script
+        '';
+    };
+  };
+  up = pkgs.writeScript "up" ''
+    #!${pkgs.bash}/bin/bash
+    PATH=${makeBinPath [pkgs.nettools pkgs.iproute]}:$PATH
+
+    LOCAL_GATEWAY=$(ip r | grep "default via" | grep -oE "([0-9]+\.[0-9]+\.[0-9]+\.[0-9]+)" | head -n1)
+    ip route del default via $LOCAL_GATEWAY
+    ip route add default dev tun0 proto static scope link metric 50
+    if [ ! -d /var/run/vpnc ]; then
+      mkdir -p /var/run/vpnc
+    fi
+    echo -ne $LOCAL_GATEWAY > /var/run/vpnc/LOCAL_GATEWAY
+  '';
+  down = pkgs.writeScript "down" ''
+    #!${pkgs.bash}/bin/bash
+    PATH=${makeBinPath [pkgs.nettools pkgs.iproute]}:$PATH
+
+    LOCAL_GATEWAY=$(cat /var/run/vpnc/LOCAL_GATEWAY)
+    ip route add default via $LOCAL_GATEWAY
+    ip route del default dev tun0
+  '';
+  buildService = key: value: nameValuePair "vpnc-${key}" {
+    description = "VPNC for ${key}";
+    # No automatic startup, password
+
+    path = [ pkgs.iptables pkgs.nettools pkgs.iproute ];
+    serviceConfig = {
+      ExecStart = "${pkgs.vpnc}/bin/vpnc /etc/vpnc/${key}.conf";
+      ExecStartPost = "${up}";
+      ExecStopPost = "${down}";
+      Type = "forking";
+    };
+  };
+
+  cfg = config.services.networking.vpnc;
+  vpnModule = types.submodule {
+    options = {
+      gateway = mkOption {
+        default = "";
+        type = types.separatedString "";
+        description = "Gateway address";
+      };
+      id = mkOption {
+        default = "linux";
+        type = types.separatedString "";
+        description = "Your group name";
+      };
+      secret = mkOption {
+        default = "";
+        type = types.separatedString "";
+        description = "Your shared secret";
```

Comment:
This shouldn't be part of the nix store, but instead a path where things are read from at runtime.

Maybe it's better to do that for the whole config file - I assume ipsec supports a lot of more options as well…

Reply:
I second that. Never store secrets in the Nix store (unless hashed). Better point to a file.